ksmt-yices/src/main/kotlin/org/ksmt/solver/yices/KYicesExprConverter.kt

```diff
@@ -0,0 +1,565 @@
+package org.ksmt.solver.yices
+
+import com.sri.yices.Constructor
+import com.sri.yices.Terms
+import com.sri.yices.Types
+import org.ksmt.KContext
+import org.ksmt.decl.KConstDecl
+import org.ksmt.expr.KExpr
+import org.ksmt.solver.util.KExprConverterBase
+import org.ksmt.decl.KDecl
+import org.ksmt.decl.KFuncDecl
+import org.ksmt.sort.*
+import org.ksmt.utils.mkConst
+import org.ksmt.utils.mkFreshConst
+import java.util.*
+
+open class KYicesExprConverter(
+    private val ctx: KContext,
+    private val yicesInternCtx: KYicesInternalizationContext
+) : KExprConverterBase<YicesTerm>() {
+    private var argTypes: Queue<KSort> = LinkedList()
+
+    fun <T : KSort> YicesTerm.convert(): KExpr<T> {
+        /**
+         * In Yices arrays are represented as functions,
+         * so before the conversion we examine the expression
+         * to distinguish between arrays and functions
+         */
+        KYicesExprPreprocessor().processExpr(this)
+        return this.convertFromNative()
+    }
+
+    private fun castToInt(expr: KExpr<KRealSort>): KExpr<KIntSort> = ctx.mkRealToInt(expr)
+    private fun castToReal(expr: KExpr<KIntSort>): KExpr<KRealSort> = ctx.mkIntToReal(expr)
+
+    private fun <D: KSort> castArrayRangeToInt(expr: KExpr<KArraySort<D, KRealSort>>): KExpr<KArraySort<D, KIntSort>> = with(ctx) {
+        val indexDecl = expr.sort.domain.mkFreshConst("index")
+
+        mkArrayLambda(indexDecl.decl, mkRealToInt(expr.select(indexDecl)))
+    }
+
+    private fun <D: KSort> castArrayRangeToReal(expr: KExpr<KArraySort<D, KIntSort>>): KExpr<KArraySort<D, KRealSort>> = with(ctx) {
+        val indexDecl = expr.sort.domain.mkFreshConst("index")
+
+        mkArrayLambda(indexDecl.decl, mkIntToReal(expr.select(indexDecl)))
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    override fun findConvertedNative(expr: YicesTerm): KExpr<*>? = with(ctx) {
+        val convertedExpr = yicesInternCtx.findConvertedExpr(expr)
+
+        /**
+         * Yices doesn't distinguish between IntSort and RealSort,
+         * therefore we need to cast terms to suitable sorts if needed
+         */
+        val argType = argTypes.poll() ?: return convertedExpr
+
+        if (convertedExpr != null && convertedExpr.sort != argType) {
+            check(convertedExpr.sort is KArraySort<*, *>
+                    && (convertedExpr.sort as KArraySort<*, *>).range is KArithSort<*>
+                    || convertedExpr.sort is KArithSort<*>) {
+                "Unexpected expression sort ${convertedExpr.sort}"
+            }
+
+            when (argType) {
+                is KIntSort -> return castToInt(convertedExpr as KExpr<KRealSort>)
```

Comment:
It's better to check here both: expected argument sort and actual expression sort.